File: private/tool/mvn/Project.scala
package mvn

import mvn.Project.ModelDirective
import org.apache.maven.model.io.xpp3.{MavenXpp3Reader, MavenXpp3Writer}
import org.apache.maven.model.{Dependency, Model}
import org.codehaus.plexus.util.dag.{DAG, TopologicalSorter}

import java.io.{File, FileInputStream, FileOutputStream}
import java.nio.file.Path
import scala.annotation.meta.{beanGetter, beanSetter}
import scala.annotation.tailrec
import scala.collection.JavaConverters._
import scala.collection.mutable
import scala.util.Try

class Project(file: File) {

  private lazy val model: Model = {
    val in = new FileInputStream(file)
    try {
      new MavenXpp3Reader().read(new FileInputStream(file))
    } finally {
      in.close()
    }
  }

  /**
   * Project ID.
   */
  lazy val id: String = file.toPath
    .toAbsolutePath.toString

  private def parentRelativePath: Option[String] = {
    Option(model.getParent).map(_.getRelativePath)
  }

  /**
   * Parent project ID.
   */
  lazy val parentId: Option[String] = for {
    aParent <- parentRelativePath
  } yield {
    file.toPath.getParent.resolve(aParent).normalize.toAbsolutePath.toString
  }

  /**
   * Parent project pom.xml file.
   */
  lazy val parentPath: Option[Path] = {
    for {
      aParent <- parentRelativePath
    } yield {
      file.toPath.getParent.resolve(aParent).normalize
    }
  }

  /**
   * Parent project.
   *
   * @return
   */
  def parentProject: Option[Project] = {
    parentPath.map(d => new Project(d.toFile))
  }

  /**
   * Emit pom file
   *
   * @return
   */
  def emitPom(dirs: ModelDirective*): Try[File] = Try {
    val p = file.getParentFile.toPath.resolve("pom.bazelizer.__gen__.xml")
    val aFile = p.toFile
    if (aFile.exists()) {
      aFile
    } else {
      val newModel = model.clone()
      dirs.foreach(_.apply(newModel))
      writePomFile(aFile, newModel)

      println(aFile)
      aFile
    }
  }

  private def writePomFile(aFile: File, newModel: Model): Unit = {
    val out = new FileOutputStream(aFile)
    try {
      new MavenXpp3Writer().write(new FileOutputStream(aFile), newModel)
    } finally {
      out.close()
    }
  }


  override def toString: String = s"Pom(${model.getId})"
}


object Project {

  //noinspection VarCouldBeVal
  private class ProjectDefStruct {
    @beanGetter
    @beanSetter
    var file: Path = _
  }

  /**
   * From json.
   *
   * @param line a json line
   * @return Project
   */
  def apply(line: String): Project = {
    val item = JSONLSupport.GSON.fromJson(line, classOf[ProjectDefStruct])
    new Project(
      item.file.toFile)
  }

  trait ModelDirective {
    def apply(m: Model): Unit
  }

  object ModelDirective {
    case class Deps(deps: Seq[Dep]) extends ModelDirective {
      override def apply(model: Model): Unit = {
        for (d <- deps) {
          val dependency = new Dependency
          dependency.setGroupId(d.gid)
          dependency.setArtifactId(d.aid)
          dependency.setVersion(d.v)
          model.addDependency(dependency)
        }
      }
    }
  }


  /**
   * Topology sort.
   *
   * @param items projects
   * @return
   */
  case class TopologySorted(private val items: Seq[Project]) extends Iterable[Project] {
    private val dag = new DAG()
    private val vertices = mutable.Map.empty[String, Project]

    for (project <- items) {
      addVertex(project)
    }

    for (project <- items) {
      project.parentId match {
        case None =>
        case Some(pId) => dag.addEdge(project.id, pId)
      }
    }

    override def iterator: Iterator[Project] = {
      val labels = TopologicalSorter.sort(dag)
      labels.asScala.map(vertices).iterator
    }

    @tailrec
    private def addVertex(aProject: Project): Unit = {
      val projectId = aProject.id
      dag.addVertex(projectId)

      if (!vertices.contains(projectId))
        vertices.update(projectId, aProject)

      aProject.parentProject match {
        case None => ()
        case Some(value) => addVertex(value)
      }
    }
  }

}

File: private/src/main/java/com/wix/incubator/mvn/Project.java
package com.wix.incubator.mvn;

import com.google.common.collect.Iterables;
import com.google.common.hash.Hashing;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import org.apache.maven.model.Dependency;
import org.apache.maven.model.Model;
import org.apache.maven.model.Parent;
import org.apache.maven.model.io.xpp3.MavenXpp3Reader;
import org.apache.maven.model.io.xpp3.MavenXpp3Writer;
import org.codehaus.plexus.util.dag.CycleDetectedException;
import org.codehaus.plexus.util.dag.DAG;
import org.codehaus.plexus.util.dag.TopologicalSorter;
import org.codehaus.plexus.util.xml.pull.XmlPullParserException;
import picocli.CommandLine;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.*;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Maven project.
 */
public class Project {

    @AllArgsConstructor
    public static class Output {
        final String src;
        final Path dest;
    }

    @AllArgsConstructor
    public static class PomFile {
        public final File file;
        public final Model model;

        public Path target() {
            return file.toPath().getParent().resolve("target").toAbsolutePath();
        }
    }

    /**
     * New project from pom file
     *
     * @param p a file
     * @return maven project
     */
    public static Project createProject(Path p) {
        return new Project(p.toFile(), Collections.emptyList());
    }

    /**
     * New projet from json description.
     *
     * @param json json str
     * @return maven project
     */
    public static Project createProject(String json) {
        final MavenProjectDTO dto = Cli.GSON.fromJson(json, MavenProjectDTO.class);
        return new Project(dto.file.toFile(), dto.flags);
    }

    private static class MavenProjectDTO {
        public Path file;
        public List<String> flags;
    }

    public static List<Project> sortProjects(List<Project> projects) throws CycleDetectedException {
        DAG dag = new DAG();
        Map<String, Project> vertices = new HashMap<>();
        for (Project project : projects) {
            addVertex(dag, vertices, project);
        }
        for (Project project : projects) {
            if (project.parentId != null) {
                dag.addEdge(project.id, project.parentId);
            }
        }
        final List<String> ids = TopologicalSorter.sort(dag);
        return ids.stream().map(vertices::get).filter(Objects::nonNull).collect(Collectors.toList());
    }

    private static void addVertex(DAG dag, Map<String, Project> vertices, Project project) {
        if (project == null) return;
        dag.addVertex(project.id);
        vertices.putIfAbsent(project.id, project);
        addVertex(dag, vertices, getParent(project).orElse(null));
    }

    private static Optional<Project> getParent(Project p) {
        return Optional.ofNullable(p.pomParentFile).map(Project::new);
    }

    private final Model model;
    private final String id;
    private final String parentId;
    private final File pomSrcFile;
    private final File pomParentFile;

    @Getter
    private final Args args;

    private Project(File file) {
        this(file, Collections.emptyList());
    }

    private Project(File file, List<String> flags) {
        try (FileInputStream is = new FileInputStream(file)) {
            MavenXpp3Reader r = new MavenXpp3Reader();
            this.model = r.read(is);
        } catch (XmlPullParserException | IOException e) {
            throw new IllegalStateException(e);
        }

        Optional<Path> parentAbsPath = Optional.ofNullable(model.getParent())
                .map(Parent::getRelativePath)
                .map(p -> file.toPath().toAbsolutePath().getParent().resolve(p).normalize());
        this.parentId = parentAbsPath.map(Path::toString).orElse(null);
        this.pomParentFile = parentAbsPath.map(Path::toFile).orElse(null);
        this.args = createArgs(this.model, flags);
        this.pomSrcFile = file;
        this.id = generateId(file, args);
    }

    private static String generateId(File file, Args args) {
        return file.toPath().toAbsolutePath() + ":" + args.toHash();
    }

    private static Project.Args createArgs(Model model, List<String> flags) {
        if (!flags.isEmpty()) {
            final Cmd.ExecutionOpts options = new Cmd.ExecutionOpts();
            final CommandLine.ParseResult result = new CommandLine(options)
                    .parseArgs(flags.toArray(new String[0]));
            if (!result.errors().isEmpty()) {
                Console.error(model, "project flags are invalid:");
                result.errors().forEach(e -> Console.error(model, " -" + e.getMessage()));
                throw new IllegalArgumentException("invalid flags for project {" + model + "}");
            }
            return Project.Args.builder()
                    .profiles(options.mavenActiveProfiles)
                    .modelVisitor(options.visitor())
                    .cmd(options.mavenArgs)
                    .build();
        } else {
            return Project.Args.builder().build();
        }
    }

    /**
     * Save pom file.
     *
     * @param args deps
     * @return file
     * @throws IOException if any
     */
    public PomFile emitPom(Project.Args args) throws IOException {
        final Path newPom = pomSrcFile.getParentFile().toPath().resolve("pom.__bazelizer__.xml");
        final Model newModel = model.clone();
        args.modelVisitor.apply(
                newModel
        );

        for (Dep dep : args.deps) {
            final Dependency d = new Dependency();
            d.setArtifactId(dep.artifactId);
            d.setGroupId(dep.groupId);
            d.setVersion(dep.version);
            d.setScope(dep.scope());
            newModel.getDependencies().add(d);
        }

        if (!newPom.toFile().exists()) {
            final MavenXpp3Writer writer = new MavenXpp3Writer();
            try (OutputStream out = Files.newOutputStream(newPom,
                    StandardOpenOption.CREATE_NEW, StandardOpenOption.TRUNCATE_EXISTING)) {
                writer.write(out, newModel);
            }
        }
        return new PomFile(newPom.toFile(), newModel);
    }

    @Override
    public String toString() {
        String suf = args.toHash();
        if (!suf.isEmpty()) suf = "/" + suf;
        return model.toString() + suf;
    }



    @AllArgsConstructor
    @Builder
    public static class Args {
        @Builder.Default
        public final List<String> cmd = Collections.emptyList();
        @Builder.Default
        public final List<Dep> deps = Collections.emptyList();
        @Builder.Default
        public final ModelVisitor modelVisitor = d -> {};
        @Builder.Default
        public final List<String> profiles = Collections.emptyList();

        public Args merge(Args other) {
            return Args.builder()
                    .cmd(Stream.concat(cmd.stream(), other.cmd.stream()).distinct().collect(Collectors.toList()))
                    .modelVisitor(modelVisitor.andThen(other.modelVisitor))
                    .deps(Stream.concat(deps.stream(), other.deps.stream()).collect(Collectors.toList()))
                    .profiles(Stream.concat(profiles.stream(), other.profiles.stream()).collect(Collectors.toList()))
                    .build();
        }

        @Override
        public String toString() {
            final String cmdStr = String.join(" ", cmd);
            final String profiles = this.profiles.stream()
                    .map(s -> "-P " + s).collect(Collectors.joining(" "));
            return "[" + cmdStr + (profiles.isEmpty() ? "" : " " + profiles) + "]";
        }

        @SuppressWarnings("UnstableApiUsage")
        public String toHash() {
            final Iterable<String> str = Iterables.concat(cmd, profiles);
            return Iterables.isEmpty(str) ? ""
                    : Hashing.murmur3_32()
                    .hashString(String.join(" ", str), StandardCharsets.UTF_8).toString();

        }
    }


    public interface ModelVisitor {
        ModelVisitor NOP = d -> {};

        void apply(Model model);

        default ModelVisitor andThen(ModelVisitor after) {
            Objects.requireNonNull(after);
            return (t) -> { apply(t); after.apply(t); };
        }
    }

    @SuppressWarnings("UnusedReturnValue")
    public static class DropAllDepsModelVisitor implements ModelVisitor {
        private Predicate<Dependency> excludeFilter = null;

        public DropAllDepsModelVisitor addIgnores(Collection<String> c) {
            c.forEach(this::addIgnore);
            return this;
        }

        public DropAllDepsModelVisitor addIgnore(String coords) {
            excludeFilter = Optional.ofNullable(excludeFilter)
                    .map(other -> other.or(matchCoordsExpr(coords)))
                    .orElseGet(() -> matchCoordsExpr(coords));
            return this;
        }

        @Override
        public void apply(Model model) {
            model.getDependencies().removeIf(d -> excludeFilter == null || !excludeFilter.test(d));
        }

        /**
         * Rule accept only by maven coordinates pattern
         *
         * @param coords pattern
         */
        private static Predicate<Dependency> matchCoordsExpr(String coords) {
            final int split = coords.indexOf(":");
            if (split == -1) throw new IllegalArgumentException(
                    "illegal expression be in format of'<artifactId|*>:<groupId|*>' ");
            String groupIdPtn = coords.substring(0, split);
            String artifactIdPtn = coords.substring(split + 1);
            final Predicate<Dependency> groupIdFilter = matchExpression(groupIdPtn, Dependency::getGroupId);
            final Predicate<Dependency> artifactIdFilter = matchExpression(artifactIdPtn, Dependency::getArtifactId);
            return groupIdFilter.and(artifactIdFilter);
        }

        private static Predicate<Dependency> matchExpression(String coords, Function<Dependency, String> fn) {
            if (coords.equals("*"))
                return d -> true;
            if (coords.startsWith("*")) {
                final int split = coords.indexOf("*");
                String suf = coords.substring(split + 1);
                return d -> fn.apply(d).endsWith(suf);
            }
            if (coords.endsWith("*")) {
                final int split = coords.indexOf("*");
                String pref = coords.substring(0, split);
                return d -> fn.apply(d).startsWith(pref);
            }
            return d -> fn.apply(d).equals(coords);
        }
    }

    @AllArgsConstructor
    public static class ChangeArtifactId implements ModelVisitor {
        private final String id;

        @Override
        public void apply(Model model) {
            model.setArtifactId(id);
        }
    }

}
